### PR TITLE
fix: improve the git shallow clone logic

### PR DESCRIPTION
**Pull request by pmuir**
#### Submitter checklist

- [ ] Change is code complete and matches issue description.
- [ ] Change is covered by existing or new tests.

#### Which issue this PR fixes

fixes #

### CODE REVIEW

**Comment by pmuir** [comment]
/hold

We need to test this extensively before we merge it!

File: cmd/codegen/util/go_test.go [hunk 1]
```diff
@@ -133,7 +133,7 @@ func Test_go_get(t *testing.T) {
 		assert.Fail(t, "unable to set env variable")
 	}
 
-	err = GoGet("github.com/golang/lint/golint", "", tmpGoDir, false, false)
+	err = GoGet("golang.org/x/lint/golint", "", tmpGoDir, false, false)
```
Comment (hferentschik) [inline cmd/codegen/util/go_test.go]: that's obsolete by now :-)

File: pkg/collector/git_collector.go [hunk 2]
```diff
@@ -2,14 +2,15 @@ package collector
...
-	"github.com/jenkins-x/jx/pkg/gits"
-	"github.com/jenkins-x/jx/pkg/log"
-	"github.com/jenkins-x/jx/pkg/util"
-	"github.com/pkg/errors"
 	"io/ioutil"
 	"os"
 	"path/filepath"
 	"strings"
+
+	"github.com/jenkins-x/jx/pkg/gits"
```
Comment (hferentschik) [inline pkg/collector/git_collector.go]: Do you have some other setting for import ordering?

File: pkg/gits/git_cli.go [hunk 3]
```diff
@@ -25,14 +26,25 @@ const (
...
+}
 
 // NewGitCLI creates a new GitCLI instance
 func NewGitCLI() *GitCLI {
 	return &GitCLI{}
 }
 
+// NewVerboseGitCLI creates a new GitCLI instance with verbose logging
+func NewVerboseGitCLI() *GitCLI {
+	return &GitCLI{
```
Comment (hferentschik) [inline pkg/gits/git_cli.go]: 👍

File: pkg/gits/git_cli.go [hunk 4]
```diff
@@ -66,97 +78,134 @@ func (g *GitCLI) FindGitConfigDir(dir string) (string, string, error) {
...
 // Clone clones the given git URL into the given directory
 func (g *GitCLI) Clone(url string, dir string) error {
-	return g.clone(dir, url, "", false, false, "", "", "")
+	return g.clone(dir, url, "", false, "", "", "")
 }
 
 // Clone clones a single branch of the given git URL into the given directory
 func (g *GitCLI) ShallowCloneBranch(url string, branch string, dir string) error {
-	return g.clone(dir, url, "", true, false, branch, "", "")
+	return g.clone(dir, url, "", true, branch, "", "")
```
Comment (hferentschik) [inline pkg/gits/git_cli.go]: Godoc wrong on this method

File: pkg/gits/git_cli.go [hunk 5]
```diff
@@ -66,97 +78,134 @@ func (g *GitCLI) FindGitConfigDir(dir string) (string, string, error) {
...
+	}
+	if commitish == "" && pullRequest == "" {
+		commitish = "master"
+	}
+
+	refspec := ""
+	treeish := ""
+	remoteBranch := ""
+	upstream := ""
+	if pullRequest != "" {
```
Comment (hferentschik) [inline pkg/gits/git_cli.go]: Could this not split out?

File: pkg/gits/git_cli_test.go [hunk 6]
```diff
@@ -37,3 +43,173 @@ func Test(t *testing.T) {
...
 }
+
+func TestGitCLI_ShallowCloneNoCommitishOrPr(t *testing.T) {
+	t.Parallel()
+	gitter := gits.NewGitCLI()
+	remoteDir := createLocalGitRepo(t, gitter)
+	localDir, err := ioutil.TempDir("", "git-clone")
+	assert.NoError(t, err)
+	// Add another commit
+	generateFileAndAddToGitRepo(t, remoteDir, "master", gitter)
```
Comment (hferentschik) [inline pkg/gits/git_cli_test.go]: 👍 for this type of test helpers. This goes into the direction I had in mind.

**Comment by hferentschik** [comment]
@pmuir what should we do with this PR. Seems it is terribly out of date atm. Is there something you want to extract into a new PR? I recommend closing this PR. WDYT?

**Comment by hferentschik** [comment]
Closing this PR for now.

**Comment by pmuir** [comment]
I've reopened it. We still need to apply it, but we agreed we need the staging env first.

**Comment by hferentschik** [comment]
> I've reopened it. We still need to apply it, but we agreed we need the staging env first.

But it is not even rebased. Given that we are not going to apply it in this form, what is the point of it being open? After all the branch and all still exist. 

I am a big proponent to keeping the list of open pull request as small as possible. It reduces the cognitive load.

**Comment by ccojocar** [comment]
@pmuir @hferentschik What's the deal with this one? Do you still plan to work on it? Shell we close it?

**Comment by ccojocar** [comment]
@pmuir @hferentschik I am closing this since it seems to be quite a bit out of sync.

**Pull request closed without merging**